lib/phoenix_live_view/channel.ex

```diff
@@ -91,19 +123,61 @@ defmodule Phoenix.LiveView.Channel do
     {:noreply, reply(%{state | components: new_components}, msg.ref, :ok, %{})}
   end
 
+  def handle_info(%Message{topic: topic, event: "progress"} = msg, %{topic: topic} = state) do
+    %{socket: socket} = state
+    %{"ref" => ref, "entry_ref" => entry_ref, "progress" => progress} = msg.payload
+    new_socket = Upload.update_progress(socket, ref, entry_ref, progress)
+
+    handle_changed(state, new_socket, msg.ref)
+  end
+
+  def handle_info(%Message{topic: topic, event: "allow_upload"} = msg, %{topic: topic} = state) do
+    %{"ref" => ref} = msg.payload
+    upload_conf = Upload.get_upload_by_ref!(state.socket, ref)
+
+    case Upload.put_entries(state.socket, upload_conf, msg.payload["entries"]) do
+      {:ok, new_socket} ->
+        case Upload.generate_preflight_response(new_socket, upload_conf.name) do
+          {:ok, uploader_reply, new_socket} ->
+            {:noreply, new_state} = handle_changed(state, new_socket, nil)
+            reply(new_state, msg.ref, :ok, uploader_reply)
+            {:noreply, new_state}
+
+          {:error, error_resp, new_socket} ->
+            {:noreply, new_state} = handle_changed(state, new_socket, nil)
+            reply(new_state, msg.ref, :ok, error_resp)
+            {:noreply, new_state}
+        end
+
+      {:error, new_socket, errors} ->
+        errors_reply =
+          Enum.map(errors, fn {ref, msg} -> %{"ref" => ref, "reason" => to_string(msg)} end)
+
+        {:noreply, new_state} = handle_changed(state, new_socket, nil)
+        reply(new_state, msg.ref, :ok, %{error: errors_reply})
+        {:noreply, state}
+    end
+  end
+
   def handle_info(%Message{topic: topic, event: "event"} = msg, %{topic: topic} = state) do
     %{"value" => raw_val, "event" => event, "type" => type} = msg.payload
     val = decode_event_type(type, raw_val)
+    new_state = maybe_update_uploads(state, msg.payload)
```

Comment:
I thought the uploads would always be sent out of band... which events can have the uploads information in them? Is it worth to have two ways of sending uploads?

Reply:
The uploads metadata is sent with the phx_change input events, so needed here